### ISSUE TITLE
trade-goods.plist capacity property not working as intended

### ISSUE THREAD

**Issue by AnotherCommander** [OoliteProject/oolite#175]
When the "capacity" key is used in trade-goods.plist, certain values of it can cause strange and unintended behaviour. 

To see the bug in action, use trunk and this trade-goods.plist inside an AddOns/Config folder without any other expansions installed:
```
{
	"testcommodity" =
	{
		"name" = "Test Good";
		//"capacity" = 13;
		"classes" = ("oolite-consumer","oolite-edible","oolite-farming");
		"quantity_unit" = 0;  // 0=t
		"peak_export" = 7; // Poor Ag
		"peak_import" = 0; // Rich Ind
		"price_average" = 50; // decicredits
		// fraction of average ~= 2.75 credits
		"price_economic" = 0.55;
		// fraction of average ~= 0.2 credits
		"price_random" = 0.04;
		"quantity_average" = 13.5; // gets rounded
		"quantity_economic" = 0.52;
		"quantity_random" = 0.04;
		"legality_export" = 0;
		"legality_import" = 0;
		"trumble_opinion" = 1.0;
		"sort_order" = 100;
	}
}    
```
All entries except the first two have been copied from "food" from the core trade-goods.plist. As you can see, capacity is intially commented out. Launch the game and do a normal start at Lave. Go straight to the market screen, you will find Test Good listed. You can try buying and selling, all good.

Now uncomment the capacity entry above and Shift-start and do the same steps as before. Now you will find that Test Good is listed again, with a quantity of 16t and you can buy, but cannot sell back. If you just press Enter on Test Good, it will instantly buy 13t of Test Good. Pressing Enter again will sell back 10t, giving 13t to the station and 3t remaining in the hold.

I tried to log what's going on and this is the result:
```
10:49:29.053 [unclassified]: good: [commodity-name alien_items], localCapacity: 7, capacity:31, q: 0, baseCapacity: 127
10:49:29.053 [unclassified]: good: [commodity-name gem_stones], localCapacity: 31, capacity:31, q: 8, baseCapacity: 127
10:49:29.053 [unclassified]: good: [commodity-name minerals], localCapacity: 31, capacity:31, q: 59, baseCapacity: 127
10:49:29.053 [unclassified]: good: [commodity-name computers], localCapacity: 7, capacity:31, q: 0, baseCapacity: 127
10:49:29.054 [unclassified]: good: [commodity-name textiles], localCapacity: 7, capacity:31, q: 16, baseCapacity: 127
10:49:29.054 [unclassified]: good: [commodity-name machinery], localCapacity: 7, capacity:31, q: 11, baseCapacity: 127
10:49:29.054 [unclassified]: good: [commodity-name slaves], localCapacity: 7, capacity:31, q: 4, baseCapacity: 127
10:49:29.054 [unclassified]: good: [commodity-name radioactives], localCapacity: 31, capacity:31, q: 18, baseCapacity: 127
10:49:29.066 [unclassified]: good: [commodity-name platinum], localCapacity: 31, capacity:31, q: 10, baseCapacity: 127
10:49:29.066 [unclassified]: good: [commodity-name luxuries], localCapacity: 3, capacity:31, q: 15, baseCapacity: 127
10:49:29.066 [unclassified]: good: [commodity-name firearms], localCapacity: 3, capacity:31, q: 0, baseCapacity: 127
10:49:29.066 [unclassified]: good: [commodity-name furs], localCapacity: 7, capacity:31, q: 50, baseCapacity: 127
10:49:29.066 [unclassified]: good: [commodity-name narcotics], localCapacity: 3, capacity:31, q: 0, baseCapacity: 127
10:49:29.066 [unclassified]: good: [commodity-name gold], localCapacity: 31, capacity:31, q: 8, baseCapacity: 127
10:49:29.066 [unclassified]: good: [commodity-name alloys], localCapacity: 3, capacity:31, q: 21, baseCapacity: 127
10:49:29.066 [unclassified]: good: [commodity-name food], localCapacity: 15, capacity:31, q: 17, baseCapacity: 127
10:49:29.066 [unclassified]: good: [commodity-name liquor_wines], localCapacity: 15, capacity:31, q: 29, baseCapacity: 127
10:49:29.066 [unclassified]: good: Test Good, localCapacity: 15, capacity:31, q: 16, baseCapacity: 13
10:49:29.066 [unclassified]: Creating test definition with price 40 and quantity 0
10:49:29.067 [unclassified]: Set good before: {capacity = 15; classes = ("oolite-consumer", "oolite-edible", "oolite-farming"); key = testcommodity; legality_export = 0; legality_import = 0; name = "Test Good"; peak_export = 7; peak_import = 0; price = 40; price_average = 50; price_economic = 0.55; price_random = 0.04; quantity = 0; quantity_average = 13.5; quantity_economic = 0.52; quantity_random = 0.04; quantity_unit = 0; sort_order = 100; trumble_opinion = 1.0; }
10:49:29.067 [unclassified]: Set good after: {capacity = 15; classes = ("oolite-consumer", "oolite-edible", "oolite-farming"); key = testcommodity; legality_export = 0; legality_import = 0; name = "Test Good"; peak_export = 7; peak_import = 0; price = 40; price_average = 50; price_economic = 0.55; price_random = 0.04; quantity = 0; quantity_average = 13.5; quantity_economic = 0.52; quantity_random = 0.04; quantity_unit = 0; sort_order = 100; trumble_opinion = 1.0; }
10:49:29.067 [unclassified]: good: [commodity-name alien_items], localCapacity: 3, capacity:31, q: 0, baseCapacity: 127
10:49:29.067 [unclassified]: good: [commodity-name gem_stones], localCapacity: 31, capacity:31, q: 8, baseCapacity: 127
10:49:29.067 [unclassified]: good: [commodity-name minerals], localCapacity: 31, capacity:31, q: 59, baseCapacity: 127
10:49:29.067 [unclassified]: good: [commodity-name computers], localCapacity: 3, capacity:31, q: 0, baseCapacity: 127
10:49:29.067 [unclassified]: good: [commodity-name textiles], localCapacity: 7, capacity:31, q: 16, baseCapacity: 127
10:49:29.067 [unclassified]: good: [commodity-name machinery], localCapacity: 7, capacity:31, q: 11, baseCapacity: 127
10:49:29.067 [unclassified]: good: [commodity-name slaves], localCapacity: 3, capacity:31, q: 4, baseCapacity: 127
10:49:29.067 [unclassified]: good: [commodity-name radioactives], localCapacity: 31, capacity:31, q: 18, baseCapacity: 127
```
Things seem to start working again for capacity values equal to or higher than 16, which happens to be the value of q in the data logged (although q seems to be going into the good definition with a value of 0; this is a result of .the adjustQuantity: byRule method).

**Comment by ocz1** [comment]
From a quick look on the matter my first guess is, that the erratic behaviour is caused by a too high amount of in-stock articles, when there is no "place" for them. (Captain Obvious strikes again)

My theory: The market creation algorithm calculates the in-stock value before/at startUp and shipExitedWitchspace doesn't concern itself with the capacity value for this commodity (assumes the default 127 TC ?), while the buying and selling algorithms do that just fine. Buying heaps of the commodity at once (pressing enter) sets a transaction value, that leaves the stockvalue at a correct value (beneath 13 TC in the above example), while trying to buy only one at one times might end up to not do. ( 16TC(before value) - 1TC(buy value) = 15TC(after value), but 15TC > 13TC(cap value) --> exception/method failed)

The market creation algorithm has to be hunted down in the source and be made to acknowledge the trade-goods.plist capacity value.

Quickfix:
If (stockValue bigger than capacityValue), than set stockValue to capacityValue;

Not-so-quickfix:
stockQuotient = capacityValue/maxMarketCapacityValue;
stockValue *= stockQuotient;
If (stockValue bigger than capacityValue), than set stockValue to capacityValue;

If my theory is right, this should kill the bug.

**Comment by cim--** [comment]
I think what's going wrong is - because nothing in the core game gets close - while it scales the quantities fine (https://github.com/OoliteProject/oolite/blob/master/src/Core/OOCommodities.m#L333 onwards) for station capacities, it doesn't check the capacity cap for the _system_ market.

https://github.com/OoliteProject/oolite/blob/master/src/Core/OOCommodities.m#L281 onwards should use the capacity limit on the good (with MAIN_SYSTEM_MARKET_LIMIT as the default), rather than always using 127.

I'll try to put a fix in later today.

**Comment by cim--** [comment]
Should be fixed in 10ca9e2 - please test to make sure I've not missed any cases.

**Comment by AnotherCommander** [comment]
From what I can see, it seems to work fine. I will close this issue as fixed for now, we can re-open if we discover any case where it might misbehave.
